Make Travis works as expected along the current BU codebase

sickpig opened the pull request:
This a minimal set of changes to make it travis works as expected with the current BU code base. As you can see with this patch applied all configured builds pass without error: 

https://travis-ci.org/sickpig/BitcoinUnlimited/builds/172682349

Things done so far are: 

- disabling java based comparison test passing `--without-comparison-tool` to the configure script. This will disable a "large reorg" test where a `bitcoinj` instance and the just compiled `bitcoind` behavior are compared to find discrepancies in the contest of such reorg. You could examine the log here: https://travis-ci.org/sickpig/BitcoinUnlimited/jobs/171336039#L3064 

- fix transaction_test.cpp: this a one liner that re initialize the min relay transaction fee to the default value just before the dust assertion. I think that the failure happen due to the fact that `minRelayTxFee` seems to be a global variable. This is the error message we got both on 32 bit linux and 32 bit win.
```
test/transaction_tests.cpp(341): error: in "transaction_tests/test_IsStandard": 
check nDustThreshold == 546 has failed [0 != 546]
```
you could check the log yourself, see https://s3.amazonaws.com/archive.travis-ci.org/jobs/172632133/log.txt and https://s3.amazonaws.com/archive.travis-ci.org/jobs/172632134/log.txt

- Disable Qt compilation for 32 bit win platform cause it takes too long to compile it, hence the build is not able to finish and to perform `make check` and qa python test suite.
   
It would be great if I could have feedbacks especially on the first two points. On the first one I have to say that such test has been disabled also on core 0.13 branch, on the second one I can't stop thinking it seems arch dependent since both of the failures are experienced on 32 bit system.

I have a more comprehensive re-factor of the travis confs and build system here: https://github.com/sickpig/BitcoinUnlimited/tree/test/travis-ci but since we are near to a release I think it's better to reduce the amount of changes at the lowest amount possible.

sickpig committed:
[travis] disable java based comparison test

sickpig committed:
[travis] Disable Qt build for i686-pc-linux-gnu platform

Due to the fact it takes too long building and that
won't let us terminate the `make check` and the
python qa tests.

sickpig committed:
[travis] Use 3 jobs for 'make check'

ftrader commented:
In my opinion for the second point, hardcoding minRelayTxFee in the test masks different behavior in the application now. While the minRelayTxFee seems to be initialized similarly as before, there is an additional algorithm involved now which modified it in the application - cf. "Xtreme Thinblocks Auto Mempool Limiter" in main.cpp which modified minRelayTxFee based on a calculated fee cutoff limited by new parameters -minlimitertxfee and -maxlimitertxfee .

sickpig commented:
@ftrader this is the exact kind of feedbacks I was referring to, thanks. You are right. I propose to change the test and set min dust to 0 since if the node is running with xthin enabled 0 is the new default value. Alternatively we could run the test with xthin disabled or run it twice, one with xthin enabled and one with xthin disabled.

That said it seems that this PR is working properly

https://travis-ci.org/BitcoinUnlimited/BitcoinUnlimited/builds/172740523

sickpig commented:
ok this is strange. 

I've tried to replicate the `transaction_test.cpp` failure locally in two ways: cross compiling 32bit binaries on a 64bit machine and compiling 32bit binaries inside a 32bit VM. The test always passed.

I've just reverted the patch and as you can see the travis builds terminate successfully, for all the platforms. (there is a timeout but it related to the python test suite).

I'm going to try to reproduce it on win32 and I'll post here the findings.

sickpig commented:
Eventually I was able to reproduce the failure. This is what you need: 

- cross compile for 32bit windows see (https://github.com/BitcoinUnlimited/BitcoinUnlimited/blob/0.12.1bu/doc/build-windows.md)
- install wine on your machine (sudo apt-get install wine or use ppa:ubuntu-wine/ppa to get a more recent version) 
- run the `make check`

this what you'll get: 
```
test/transaction_tests.cpp(341): error: in "transaction_tests/test_IsStandard": check nDustThreshold == 546 has failed [0 != 546]
```
so it is platform specific, it happens only on win both 32 and 64 bit. Previously I said win and linux 32bit, I was wrong cause I mixed up log files.

sickpig commented:
update on 1st point (disable reorg test). This is supposed to be disabled by default (see: core 47b9374e39a00d6e981aa1f4d0175d9426d91ac8 commit), but for some reason this does not happen when cross compiling for win 32/64 bit platform. 

So it seems that we have 2 issues with win platforms: the `transaction_test.cpp` failure and the execution of a test that is supposed to be disabled. There's definitely something wrong in the setup we are currently using for the windows cross building and testing.

sickpig commented:
`transaction_test.cpp` fails only when executed together with all the other tests via `make check`, if you run it in isolation like this: 
```
﻿⁠⁠⁠⁠src/test/test_bitcoin.exe --run_test=transaction_tests﻿⁠⁠⁠⁠
```
the test pass without error. so it seems that some prev test initialized `minRelayTxFee` to zero. I think that the long term solutions is reducing the use of global variables in the code base.

In the near term I think that reinit `minRelayTxFee` to its default value is a "proper" fix.

ftrader commented:
Separate from @sickpig's proposal to temporarily disable the failing test with a TODO (which I agree with, I think it's more important to get to cleanly passing Travis tests for the rest in a first step), I had another idea for which I'll make a separate PR:

Commenting tests out (even with a TODO comment) still makes them disappear off the radar, which reduces the likelihood that they'll be fixed sooner.

I propose to enhance `rpc_tests.py` to mark certain tests as DISABLED / SKIP (on certain platforms) etc. and have that be part of the test output.

Most other unit test frameworks have a way of handling skipped tests/expected failures.
Seeing the skipped/disabled tests in the test output would at least remind that we need to fix some issues related to those tests.

sickpig committed:
[travis] Comment out Dust threshold tests.

This test pass if executed in isolation whereas it could fails
when it run together with all others due to the fact that
minRelayTxFee values could be changed by some test executed
before this one. This is happen because mineRelayTxFee is a global
variable and it is shared by all tests belonging to the suite.

TODO: add a unit test to exercise the {min,max}limitertxfee
machinery

sickpig commented:
PR updated as per my prev comment. 

@ftrader I think that having a tool to mark and report a test as skipped/disabled on a particular platform would be an awesome things to have, both for the `rpc_tests.py` and for the tests based on the boost unit test framework, like the one we're dealing with in th this PR.

sickpig commented:
Travis back to all green, I'm going to squash the commits a bit and mark the PR ready to be committed.

sickpig commented:
Squashed and ready for the committer.

ftrader commented:
@sickpig: PR#150 adds disable/skip functionality for the qa tests.

https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/150

sickpig commented:
@ftrader you rock
